### PR TITLE
Movable UI Elements!

### PR DESCRIPTION
**Pull request by RemieRichards**
Adds 2 types of Movable UI Element (Fancier name for "Screen object")
- <B>Movable:</B> Places it's centre where the cursor is
- <B>Snap:</B> Snaps itself to the nearest turf

Also includes 2 debug verbs (requiring R_DEBUG Admin rank) to spawn a Movable and a Snap UI Element. These debug verbs request a screen_loc from the user. the message recommends the numerical format for screen_loc but will happily accept the keywords instead (CENTER,CENTER)

<B>Example Gif:</B> (Note, Any unresponsiveness in the gif is the result of my mouse and gifcam hating each other)

movableuielements

### CODE REVIEW

**Comment by Deantwo** [comment]
Interesting.
So we can get better HUD icons now or what?

Not 100% sure what it is your trying to explain.

**Comment by RemieRichards** [comment]
@Deantwo This opens up the possibility for non-fixed UI Elements, E.g. if you don't like your internals icon on the right, you can just drag it off to the left (Note: That's the idea, nothing like that has been implemented yet, just the working types)

**Comment by Razharas** [comment]
Maybe its the beginning of the great interface rewrite and in the end we will get non-horrible non copypaste interfaces to any thing ever needed
I really hope so

**Comment by JJRcop** [comment]
Maybe we can get rid of browse/nanoui and move to complete object based uis

**Comment by Deantwo** [comment]
That sounds awesome!

**Comment by Chocobro** [comment]
:+1: this feels really good. it's a good clunk.

File: code/_onclick/hud/movable_screen_objects.dm [hunk 1]
```diff
@@ -0,0 +1,42 @@
+
+//////////////////////////
+//Movable Screen Objects//
+//   By RemieRichards	//
+//////////////////////////
```
Comment (MrPerson) [inline code/_onclick/hud/movable_screen_objects.dm]: The evolved form of // Movable screen objects- RR

**Comment by An0n3** [comment]
Sounds like a neat feature.  Lots of potential.

**Comment by paprka** [comment]
This is very cool

**Comment by RemieRichards** [comment]
My plans for when this is merged are as follows:
* Add functionality to allow users to change their screen objects' positions (Using a sort of "HUD Lock" setting so you don't accidentally drag things during play
* Port appropriate screen objects on the UI to /snap (Snap is a tad more intuitive than /movable)
* Have a look at refactoring the addition/removal of HUD Objects so that they can possibly be saved/loaded in their custom locations.

**Comment by MrPerson** [comment]
Can you add a debug button so we can spawn a bunch of testing button on screen so we can try this stuff out on a live server?

**Comment by RemieRichards** [comment]
@MrPerson what I used for my tests was literally a big M (movable) and a big S (Snap) objects and dragging them around.

I assume you'll want this as an admin only verb?

**Comment by RemieRichards** [comment]
Created some debug verbs for R_DEBUG Rank Admins (These are even nicer then the ones I used to originally test this, I'm that generous!)

**Comment by Deantwo** [comment]
Not even a nice picture for us sad people that can't test it ourselvs right now?
Because it still sound awesome.

**Comment by RemieRichards** [comment]
@Deantwo if you'll read the OP I mention it's incredibly annoying for me to produce any for of digital representation that would properly show this. just a still picture would be nothing new.

**Comment by Deantwo** [comment]
Yeah sorry, but there you only mentioned gif. ._.
So nevermind then, sorry again.

**Comment by Jordie0608** [comment]
Changelog.

**Comment by RemieRichards** [comment]
I don't agree that this needs a changelog. This doesn't affect players at the moment, and anyone interested in the use of this would have looked at the PR.

**Comment by Jordie0608** [comment]
Ah alright, I didn't realise this was framework only.

**Comment by phil235** [comment]
Tested it, it's great. But it runtimes if you click&drag the UI element to outside the screen (in the black area outside the screen).

**Comment by RemieRichards** [comment]
it runtimes? well, technically that's Byond's fault (I'm assuming the click values for screen-loc become null, I'll work on something to constrain it to the view box.

**Comment by RemieRichards** [comment]
Ok, got it sorted, fix incoming.

**Comment by GunHog** [comment]
I <3 this RR! Are you going to be making any live interfaces from this soon, or providing a guide for coding them ourselves?

**Comment by RemieRichards** [comment]
@GunHog
well it's fairly simple, take the screen object you want to make movable, and stick /movable after the /screen in it's path. if you want it to be a /snap, please add /movable/snap after the /screen in it's path.

So while producing them externally is really simple, I'd like to take a stab at doing some of it myself yes, a couple of posts up I made a little list of some things I'd like to do after this framework is merged.

**Pull request merged**